src/query/storage/m3/dynamic_cluster.go

```diff
@@ -326,13 +315,27 @@ func (d *dynamicCluster) updateClusterNamespacesWithLock() {
 
 	var (
 		newNamespaces            = make(ClusterNamespaces, 0, nsCount)
+		newNonReadyNamespaces    = make(ClusterNamespaces, 0, nsCount)
 		newAggregatedNamespaces  = make(map[RetentionResolution]ClusterNamespace)
 		newUnaggregatedNamespace ClusterNamespace
 	)
 
 	for _, nsMap := range d.namespacesByEtcdCluster {
-		for _, clusterNamespaces := range nsMap.metadataToClusterNamespaces {
+		for md, clusterNamespaces := range nsMap.metadataToClusterNamespaces {
 			for _, clusterNamespace := range clusterNamespaces {
+				status := md.Options().StagingState().Status()
+				// Don't make non-ready namespaces available for read/write in coordinator, but track
+				// them so that we can have the DB session available when we need to check their
+				// readiness in the /namespace/ready check.
+				if status != namespace.ReadyStagingStatus {
```

Comment:
is the unknown state intentionally put in here as well?

Reply:
Yup. Namespaces created before staging status was added will be `unknown`, but it's OK to promote them to `ready` all the same.